File: ios/StillMoment/Presentation/Views/GuidedMeditations/TrimEditor/PlaybackRangeCard.swift
//
//  PlaybackRangeCard.swift
//  Still Moment
//
//  Presentation Layer — "Wiedergabe-Bereich" form card in the edit sheet (shared-107).
//

import SwiftUI

/// Tappable card in the meditation edit sheet that summarizes the current playback range
/// and opens the full-screen waveform trim editor.
///
/// Two states follow the design handoff "View 1":
/// - **Untrimmed:** "Ganze Datei · {fileDuration}" + a "Bereich wählen" affordance.
/// - **Trimmed:** a static mini waveform with the selected range highlighted, the time
///   range, the audible duration, and a separate "Zuschnitt entfernen" text link.
///
/// The whole card opens the editor (`onOpenEditor`); the remove link is its own tap target
/// and resets the trim without opening the editor (`onRemoveTrim`).
struct PlaybackRangeCard: View {
    // MARK: Internal

    let fileDuration: TimeInterval
    /// Pending start offset in seconds (nil = no trim); reflects the edit sheet's uncommitted state.
    let trimStart: TimeInterval?
    /// Pending end offset in seconds (nil = no trim).
    let trimEnd: TimeInterval?
    /// Optional precomputed waveform for the mini display; nil renders the trimmed row without bars.
    let waveform: MeditationWaveform?
    let onOpenEditor: () -> Void
    let onRemoveTrim: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            self.card
            if self.isTrimmed {
                self.removeLink
            }
        }
    }

    // MARK: Private

    @Environment(\.themeColors)
    private var theme

    /// A range is "trimmed" once at least one explicit point is set.
    private var isTrimmed: Bool {
        self.trimStart != nil || self.trimEnd != nil
    }

    private var effectiveStart: TimeInterval {
        self.trimStart ?? 0
    }

    private var effectiveEnd: TimeInterval {
        self.trimEnd ?? self.fileDuration
    }

    private var audibleDuration: TimeInterval {
        max(self.effectiveEnd - self.effectiveStart, 0)
    }

    private var card: some View {
        Button(action: self.onOpenEditor) {
            VStack(alignment: .leading, spacing: 12) {
                self.header
                if self.isTrimmed {
                    self.trimmedContent
                } else {
                    self.untrimmedContent
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: 24)
                    .fill(self.theme.cardBackground.opacity(.opacitySecondary))
            )
            .overlay(
                RoundedRectangle(cornerRadius: 24)
                    .strokeBorder(self.theme.cardBorder, lineWidth: 0.5)
            )
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text("playback_range.label"))
        .accessibilityValue(Text(self.accessibilityValue))
        .accessibilityHint(Text("playback_range.a11y.openHint"))
        .accessibilityAddTraits(.isButton)
        .accessibilityIdentifier("editSheet.card.playbackRange")
    }

    private var header: some View {
        HStack {
            Text("playback_range.label")
                .textStyle(.eyebrow, color: \.textSecondary)
            Spacer()
            Image(systemName: "chevron.right")
                .font(.system(size: 14, weight: .semibold))
                .foregroundColor(self.theme.textSecondary)
                .opacity(.opacitySecondary)
        }
    }

    private var untrimmedContent: some View {
        HStack(alignment: .firstTextBaseline) {
            Text(
                String(
                    format: NSLocalizedString("playback_range.wholeFile", comment: ""),
                    EditSheetState.formatTime(self.fileDuration)
                )
            )
            .textStyle(.body, monospacedDigits: true, color: \.textPrimary)
            Spacer(minLength: 12)
            HStack(spacing: 6) {
                Text("playback_range.choose")
                    .textStyle(.caption, color: \.interactive)
                Image(systemName: "scissors")
                    .font(.system(size: 14, weight: .regular))
                    .foregroundColor(self.theme.interactive)
            }
        }
    }

    private var trimmedContent: some View {
        VStack(alignment: .leading, spacing: 10) {
            TrimWaveformView(
                waveform: self.waveform,
                isLoading: false,
                loadFailed: self.waveform == nil,
                duration: self.fileDuration,
                start: self.effectiveStart,
                end: self.effectiveEnd,
                playheadTime: nil,
                height: Self.miniWaveformHeight
            )

            HStack(alignment: .firstTextBaseline) {
                Text(
                    String(
                        format: NSLocalizedString("playback_range.range", comment: ""),
                        EditSheetState.formatTime(self.effectiveStart),
                        EditSheetState.formatTime(self.effectiveEnd)
                    )
                )
                .textStyle(.title, monospacedDigits: true, color: \.interactive)
                Spacer(minLength: 12)
                Text(
                    String(
                        format: NSLocalizedString("playback_range.audible", comment: ""),
                        EditSheetState.formatTime(self.audibleDuration)
                    )
                )
                .textStyle(.caption, monospacedDigits: true, color: \.textSecondary)
            }
        }
    }

    private var removeLink: some View {
        Button(action: self.onRemoveTrim) {
            Text("playback_range.remove")
                .textStyle(.caption, color: \.textSecondary)
                .padding(.vertical, 8)
                .padding(.horizontal, 4)
                .frame(minHeight: 44, alignment: .leading)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel(Text("playback_range.remove"))
        .accessibilityHint(Text("playback_range.a11y.removeHint"))
        .accessibilityIdentifier("editSheet.button.removeTrim")
    }

    private var accessibilityValue: String {
        guard self.isTrimmed else {
            return String(
                format: NSLocalizedString("playback_range.wholeFile", comment: ""),
                EditSheetState.formatTime(self.fileDuration)
            )
        }
        return String(
            format: NSLocalizedString("playback_range.range", comment: ""),
            EditSheetState.formatTime(self.effectiveStart),
            EditSheetState.formatTime(self.effectiveEnd)
        )
    }

    private static let miniWaveformHeight: CGFloat = 44
}

// MARK: - Previews

#if DEBUG
private let cardPreviewWaveform = MeditationWaveform(
    samples: (0..<MeditationWaveform.sampleCount).map { index in
        let fraction = Double(index) / Double(MeditationWaveform.sampleCount)
        let speech = abs(sin(fraction * 40)) * 0.8 + 0.15
        let isSilence = fraction > 0.18 && fraction < 0.82
        return Float(isSilence ? 0.06 : speech)
    }
)

#Preview("Untrimmed") {
    ThemeRootView {
        Form {
            Section {
                PlaybackRangeCard(
                    fileDuration: 1145,
                    trimStart: nil,
                    trimEnd: nil,
                    waveform: cardPreviewWaveform,
                    onOpenEditor: {},
                    onRemoveTrim: {}
                )
            } footer: {
                Text("playback_range.help")
            }
        }
        .scrollContentBackground(.hidden)
    }
}

#Preview("Trimmed") {
    ThemeRootView {
        Form {
            Section {
                PlaybackRangeCard(
                    fileDuration: 1145,
                    trimStart: 84,
                    trimEnd: 1110,
                    waveform: cardPreviewWaveform,
                    onOpenEditor: {},
                    onRemoveTrim: {}
                )
            } footer: {
                Text("playback_range.help")
            }
        }
        .scrollContentBackground(.hidden)
    }
}

#Preview("Trimmed — No Waveform") {
    ThemeRootView {
        Form {
            Section {
                PlaybackRangeCard(
                    fileDuration: 1145,
                    trimStart: 84,
                    trimEnd: 1110,
                    waveform: nil,
                    onOpenEditor: {},
                    onRemoveTrim: {}
                )
            } footer: {
                Text("playback_range.help")
            }
        }
        .scrollContentBackground(.hidden)
    }
}
#endif

File: android/app/src/main/kotlin/com/stillmoment/presentation/ui/meditations/components/PlaybackRangeCard.kt
package com.stillmoment.presentation.ui.meditations.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.ContentCut
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.stillmoment.R
import com.stillmoment.domain.models.MeditationWaveform
import com.stillmoment.presentation.ui.meditations.trimeditor.MINI_BAR_COUNT
import com.stillmoment.presentation.ui.meditations.trimeditor.MINI_HEIGHT
import com.stillmoment.presentation.ui.meditations.trimeditor.TrimWaveformSpec
import com.stillmoment.presentation.ui.meditations.trimeditor.TrimWaveformView
import com.stillmoment.presentation.ui.meditations.trimeditor.formatTrimTime
import com.stillmoment.presentation.ui.theme.LocalStillMomentColors
import com.stillmoment.presentation.ui.theme.TextStyle
import com.stillmoment.presentation.ui.theme.toComposeTextStyle

/**
 * Tappable card in the meditation edit sheet summarising the current playback range and
 * opening the full-screen trim editor (shared-107).
 *
 * Two states:
 * - **Untrimmed:** "Ganze Datei · {fileDuration}" + a "Bereich wählen" affordance.
 * - **Trimmed:** a static mini waveform with the selected range highlighted, the time range,
 *   the audible duration, and a separate "Zuschnitt entfernen" text link.
 *
 * The whole card opens the editor ([onOpenEditor]); the remove link is its own tap target
 * and resets the trim without opening the editor ([onRemoveTrim]). All values in ms.
 * 1:1 port of iOS `PlaybackRangeCard`.
 */
@Composable
fun PlaybackRangeCard(
    fileDurationMs: Long,
    trimStartMs: Long?,
    trimEndMs: Long?,
    waveform: MeditationWaveform?,
    onOpenEditor: () -> Unit,
    onRemoveTrim: () -> Unit,
    modifier: Modifier = Modifier
) {
    val isTrimmed = trimStartMs != null || trimEndMs != null
    val effectiveStart = trimStartMs ?: 0L
    val effectiveEnd = trimEndMs ?: fileDurationMs

    Column(modifier = modifier.fillMaxWidth()) {
        RangeCardBody(
            isTrimmed = isTrimmed,
            fileDurationMs = fileDurationMs,
            effectiveStartMs = effectiveStart,
            effectiveEndMs = effectiveEnd,
            waveform = waveform,
            onOpenEditor = onOpenEditor
        )
        if (isTrimmed) {
            RemoveTrimLink(onRemoveTrim = onRemoveTrim)
        }
    }
}

@Composable
private fun RangeCardBody(
    isTrimmed: Boolean,
    fileDurationMs: Long,
    effectiveStartMs: Long,
    effectiveEndMs: Long,
    waveform: MeditationWaveform?,
    onOpenEditor: () -> Unit
) {
    val theme = LocalStillMomentColors.current
    val shape = RoundedCornerShape(24.dp)
    val openHint = stringResource(R.string.playback_range_a11y_open_hint)
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(shape)
            .background(theme.cardBackground)
            .border(0.5.dp, theme.cardBorder, shape)
            .clickable { onOpenEditor() }
            .padding(horizontal = 16.dp, vertical = 14.dp)
            .semantics { contentDescription = openHint },
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
            Icon(
                Icons.AutoMirrored.Filled.KeyboardArrowRight,
                contentDescription = null,
                tint = theme.textPrimary.copy(alpha = 0.5f)
            )
        }
        if (isTrimmed) {
            TrimmedContent(fileDurationMs, effectiveStartMs, effectiveEndMs, waveform)
        } else {
            UntrimmedContent(fileDurationMs)
        }
    }
}

@Composable
private fun UntrimmedContent(fileDurationMs: Long) {
    val theme = LocalStillMomentColors.current
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = stringResource(R.string.playback_range_whole_file, formatTrimTime(fileDurationMs)),
            style = TextStyle.body.toComposeTextStyle(),
            color = theme.textPrimary
        )
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = stringResource(R.string.playback_range_choose),
                style = TextStyle.caption.toComposeTextStyle(),
                color = theme.interactive
            )
            Icon(Icons.Filled.ContentCut, contentDescription = null, tint = theme.interactive)
        }
    }
}

@Composable
private fun TrimmedContent(
    fileDurationMs: Long,
    effectiveStartMs: Long,
    effectiveEndMs: Long,
    waveform: MeditationWaveform?
) {
    val theme = LocalStillMomentColors.current
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        TrimWaveformView(
            spec = TrimWaveformSpec(
                waveform = waveform,
                durationMs = fileDurationMs,
                startMs = effectiveStartMs,
                endMs = effectiveEndMs,
                playheadTimeMs = null,
                window = 0L..fileDurationMs.coerceAtLeast(0L),
                barCount = MINI_BAR_COUNT
            ),
            isLoading = false,
            loadFailed = waveform == null,
            height = MINI_HEIGHT
        )
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = stringResource(
                    R.string.playback_range_range,
                    formatTrimTime(effectiveStartMs),
                    formatTrimTime(effectiveEndMs)
                ),
                style = TextStyle.title.toComposeTextStyle(),
                color = theme.interactive
            )
            Text(
                text = stringResource(
                    R.string.playback_range_audible,
                    formatTrimTime((effectiveEndMs - effectiveStartMs).coerceAtLeast(0L))
                ),
                style = TextStyle.caption.toComposeTextStyle(),
                color = theme.textPrimary.copy(alpha = 0.6f)
            )
        }
    }
}

@Composable
private fun RemoveTrimLink(onRemoveTrim: () -> Unit) {
    val theme = LocalStillMomentColors.current
    val hint = stringResource(R.string.playback_range_a11y_remove_hint)
    Text(
        text = stringResource(R.string.playback_range_remove),
        style = TextStyle.caption.toComposeTextStyle(),
        color = theme.textPrimary.copy(alpha = 0.6f),
        modifier = Modifier
            .clickable { onRemoveTrim() }
            .padding(vertical = 8.dp, horizontal = 4.dp)
            .semantics { contentDescription = hint }
            .testTag("editSheet.button.removeTrim")
    )
}
